### PR TITLE
Fix Windows Ruby Debug Build Issues

### PR DESCRIPTION
**Pull request by lefticus**
## Fix Debug Build Issues

I made this PR as tiny as possible. 

TL;DR

- Not a linking issue
- We never properly build Debug Ruby Windows
- Bug in Ruby
- Build script fixed, Ruby upgraded

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -357,8 +357,8 @@ if( BUILD_RUBY )
     # brew install autoconf automake libtool
     # brew link autoconf automake libtool
     ExternalProject_Add(Ruby
-      URL https://codeload.github.com/ruby/ruby/tar.gz/v2_5_1
-      URL_MD5 98b550a05fec0f7a43a22d6c37404fd4
+      URL https://codeload.github.com/ruby/ruby/tar.gz/v2_5_5
+      URL_HASH SHA256=A3BA7160C7E0CE03865CDABA9B7C26974A8A0FC23FC953269E0463C5ACAFB7A1
```
Comment (macumber) [inline CMakeLists.txt]: @lefticus can't we keep our md5s?

File: dependencies/ruby/build_ruby.bat.in [hunk 2]
```diff
@@ -15,8 +15,8 @@ echo "***** Building Ruby Release Build"
 echo "****"
 
 cd ..\Ruby-release
-set CL = /MP
-set CFLAGS = MD -Zi -W2 -wd4996 -we4028 -we4142 -O2sy-  -Zm600 
+set CL=/MP
+set CFLAGS=-MD -Zi -W2 -wd4996 -we4028 -we4142 -O2sy-  -Zm600
```
Comment (macumber) [inline dependencies/ruby/build_ruby.bat.in]: We should put a comment in here that no spaces are allowed after the `=` sign

File: CMakeLists.txt [hunk 3]
```diff
@@ -357,8 +357,8 @@ if( BUILD_RUBY )
     # brew install autoconf automake libtool
     # brew link autoconf automake libtool
     ExternalProject_Add(Ruby
-      URL https://codeload.github.com/ruby/ruby/tar.gz/v2_5_1
-      URL_MD5 98b550a05fec0f7a43a22d6c37404fd4
+      URL https://codeload.github.com/ruby/ruby/tar.gz/v2_5_5
```
Comment (macumber) [inline CMakeLists.txt]: Ubuntu 18.04 is on [Ruby 2.5.1](https://packages.ubuntu.com/bionic/ruby/ruby2.5) will this version be compatible?  Is it possible to patch 2.5.1 with the specific issue code?  Would that be safer or not?

Comment (macumber) [reply CMakeLists.txt]: SketchUp 2019 is also on Ruby 2.5.1.  The CLI should have no issues, I am only concerned about the shared library openstudio.so when loaded by Ruby 2.5.1 (either system or in SketchUp).  Should be easy to test:

```
ruby -I/path/to/openstudio.so/dir/ -e "require 'openstudio'" -e "puts OpenStudio::Model::Model.new"
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -418,12 +418,13 @@ if( BUILD_RUBY )
...
       DEPENDS ZLib FFI
-      URL https://codeload.github.com/ruby/ruby/tar.gz/v2_5_1
-      URL_MD5 98b550a05fec0f7a43a22d6c37404fd4
+      URL https://codeload.github.com/ruby/ruby/tar.gz/v2_5_5
+      URL_HASH SHA256=A3BA7160C7E0CE03865CDABA9B7C26974A8A0FC23FC953269E0463C5ACAFB7A1
       BUILD_IN_SOURCE 1
       # DLM: ruby_2_5_1_msvc2017_static_v3.tar.gz on s3 was built with experimental Ruby.win.wopen.patch patch from ruby_patches branch, not using it for now
       #PATCH_COMMAND ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/dependencies/ruby/Ruby.patch && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/dependencies/ruby/Ruby.win.patch && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/dependencies/ruby/Ruby.win.wopen.patch
-      PATCH_COMMAND ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/dependencies/ruby/Ruby.patch && ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/dependencies/ruby/Ruby.win.patch && ${PATCH_EXE} 
+      PATCH_COMMAND ${PATCH_EXE} -p1 < ${CMAKE_SOURCE_DIR}/dependencies/ruby/Ruby.patch && ${PATCH_EXE}  -p1 < ${CMAKE_SOURCE_DIR}/dependencies/ruby/Ruby.win.patch && ${PATCH_EXE}
```
Comment (macumber) [inline CMakeLists.txt]: Did something important change on this line?

**Comment by macumber** [comment]
Awesome job @lefticus !

**Comment by kbenne** [comment]
Sweet!

File: CMakeLists.txt (hunk 4 above)
Comment (lefticus) [reply CMakeLists.txt]: No, and that shouldn't have been a commit, I tried to make sure that minor (accidental) diff didn't show up in the diffs.

File: CMakeLists.txt (hunk 1 above)
Comment (lefticus) [reply CMakeLists.txt]: I didn't have an MD5 checksum tool installed on Windows, and MD5 is pretty well known to be broken at this point, so I just used SHA256. I can change it back if it matters tho.

File: dependencies/ruby/build_ruby.bat.in (hunk 2 above)
Comment (lefticus) [reply dependencies/ruby/build_ruby.bat.in]: Good call

File: CMakeLists.txt (hunk 3 above)
Comment (lefticus) [reply CMakeLists.txt]: I didn't consider the system version of ruby here on Unix, I updated the version on both sides so they would be the same, but the fixes we need are Windows specific. IMO it's best to just roll back the Unix ones in this case to minimize issues.

Comment (lefticus) [reply CMakeLists.txt]: For all of these question though, Ruby considers the API version to be 2.5.0, as we've seen in how folders are named when Ruby is installed, etc. So, as much as I trust Ruby I trust that the ABI/API is compatible and we'll have no problems with 2.5.x line.

File: CMakeLists.txt (hunk 1 above)
Comment (lefticus) [reply CMakeLists.txt]: Just noticed that SHA256 is also consistent with the other dependencies in this file.

File: CMakeLists.txt (hunk 3 above)
Comment (lefticus) [reply CMakeLists.txt]: I reverted Unix back to 2.5.1 so there's no question here about Ubuntu.

File: CMakeLists.txt (hunk 4 above)
Comment (lefticus) [reply CMakeLists.txt]: Removed empty space.

File: CMakeLists.txt (hunk 3 above)
Comment (lefticus) [reply CMakeLists.txt]: > 
> 
> SketchUp 2019 is also on Ruby 2.5.1. The CLI should have no issues, I am only concerned about the shared library openstudio.so when loaded by Ruby 2.5.1 (either system or in SketchUp). Should be easy to test:
> 
> ```
> ruby -I/path/to/openstudio.so/dir/ -e "require 'openstudio'" -e "puts OpenStudio::Model::Model.new"
> ```

Verified as loading and executing with a system installed 2.5.1 on Windows with my 2.5.5 compiled `.so`.

File: dependencies/ruby/build_ruby.bat.in (hunk 2 above)
Comment (lefticus) [reply dependencies/ruby/build_ruby.bat.in]: Done.

File: CMakeLists.txt (hunk 1 above)
Comment (lefticus) [reply CMakeLists.txt]: @macumber personally I would stick with sha256 for the reasons mentioned above, for the things that *needed* to get changed. To minimize changes and risk I did revert to 2.5.1 on Unix.

**Pull request merged**